### ISSUE TITLE
webhook 500 errors

### ISSUE THREAD

**Issue by jr0dd** [metallb/metallb#1481]
## environment

MetalLB `v0.13.2`
Kubernetes `v1.24.2`
CNI `cilium v1.11.6`
kube-proxy `not deployed`

## The bug itself, as detailed as you can.

After the initial install it took multiple attempts to apply the CR's to the cluster. The validating webhook was returning error 500 on the attempts and then finally accepted them. This greatly affects clusters managed through GitOps as reconciliations are failing.

`k apply -f cluster/apps/networking/metallb/custom-resources/bgp-advertisement.yaml`
``` 
Warning: resource bgpadvertisements/bgp-lb is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.
Error from server (InternalError): error when applying patch:
{"metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"metallb.io/v1beta1\",\"kind\":\"BGPAdvertisement\",\"metadata\":{\"annotations\":{},\"name\":\"bgp-lb\",\"namespace\":\"networking\"},\"spec\":{\"ipAddressPools\":[\"bgp-pool\"],\"peers\":[\"mikrotik\"]}}\n"}}}
to:
Resource: "metallb.io/v1beta1, Resource=bgpadvertisements", GroupVersionKind: "metallb.io/v1beta1, Kind=BGPAdvertisement"
Name: "bgp-lb", Namespace: "networking"
for: "cluster/apps/networking/metallb/custom-resources/bgp-advertisement.yaml": Internal error occurred: failed calling webhook "bgpadvertisementvalidationwebhook.metallb.io": failed to call webhook: Post "https://metallb-webhook-service.networking.svc:443/validate-metallb-io-v1beta1-bgpadvertisement?timeout=10s": proxy error from 127.0.0.1:6443 while dialing 172.20.0.186:9443, code 500: 500 Internal Server Error
```

### controller errors with info logging
```
{"level":"error","ts":1657129516.5686264,"msg":"Reconciler error","controller":"cert-rotator","object":{"name":"webhook-server-cert","namespace":"networking"},"namespace":"networking","name":"webhook-server-cert","reconcileID":"1a97259f-b252-4af3-8ac0-c727359e2278","error":"Operation cannot be fulfilled on customresourcedefinitions.apiextensions.k8s.io \"bgppeers.metallb.io\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:234"}
```

### controller errors with debug logging

```
{"action":"update","caller":"bgpadvertisement_webhook.go:70","level":"debug","name":"bgp-lb","namespace":"networking","ts":"2022-07-06T19:58:51Z","webhook":"bgpadvertisement"}
{"action":"update","caller":"bgppeer_webhook.go:60","level":"debug","name":"mikrotik","namespace":"networking","ts":"2022-07-06T19:58:51Z","webhook":"bgppeer"}
{"action":"update","caller":"ipaddresspool_webhook.go:64","level":"debug","name":"bgp-pool","namespace":"networking","ts":"2022-07-06T19:58:51Z","webhook":"ipaddresspool"}
```

### logs flooded with this community error
```
E0706 19:58:53.147844       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.24.0/tools/cache/reflector.go:167: Failed to watch *v1beta1.Community: unknown (get communities.metallb.io)
{"action":"update","caller":"bgppeer_webhook.go:60","level":"debug","name":"mikrotik","namespace":"networking","ts":"2022-07-06T19:59:15Z","webhook":"bgppeer"}
E0706 19:59:40.601150       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.24.0/tools/cache/reflector.go:167: Failed to watch *v1beta1.Community: unknown (get communities.metallb.io)
E0706 20:00:14.243217       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.24.0/tools/cache/reflector.go:167: Failed to watch *v1beta1.Community: unknown (get communities.metallb.io)
E0706 20:00:50.974510       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.24.0/tools/cache/reflector.go:167: Failed to watch *v1beta1.Community: unknown (get communities.metallb.io)
```

**Comment by fedepaol** [comment]
Just to make sure (you kind of said that in https://github.com/metallb/metallb/issues/1473#issuecomment-1176517952 ), you are deploying via the helm chart right?

**Comment by jr0dd** [comment]
Correct. It doesn't even look like  it's getting past the webhook to reach the controller the more I watch this. Ofc I could get around this by doing a patch delete with kustomize, but something is off. At least one other guy in the k8s@home discord had the same experience.

**Comment by fedepaol** [comment]
Another thing that could help are the logs of the apiserver, hoping it says a bit more than just "500"

**Comment by fedepaol** [comment]
also also (and then I'll go to bed), can you try a quick fix where you add the controller watch permisisons here https://github.com/metallb/metallb/blob/1fc030d4541eb5299025dbd5e8b8773e1b7b6570/charts/metallb/templates/rbac.yaml#L127 ?

**Comment by fedepaol** [comment]
> > also also (and then I'll go to bed), can you try a quick fix where you add the controller watch permisisons here https://github.com/metallb/metallb/blob/1fc030d4541eb5299025dbd5e8b8773e1b7b6570/charts/metallb/templates/rbac.yaml#L127
> >  ?
> 
> Yeah I'll apply that and report back. Btw this is running k3s and most things are embedded, but a checked the journalctl logs and it matched up with what I put in the first post.

I am pretty sure the community error is not relevant to the webhook, but you never know..

**Comment by jr0dd** [comment]
> 

that got rid of the community spam, but not webhook errors

**Comment by markussiebert** [comment]
I use the helm chart and have the same issue since yesterday. And metallb stoped working (not assigning L2 IP anymore)

**Comment by nick-oconnor** [comment]
@fedepaol unsure if this is related, but the webook service is also missing selectors, i.e. https://github.com/metallb/metallb/blob/2f62620c2b9218f63901f5aaab4d59440f9c2622/charts/metallb/templates/webhooks.yaml#L159 appears to be missing `metallb.selectorLabels`. If users happen to have other pods deployed to the same namespace with the label `app.kubernetes.io/component: controller`, they'll be mistakenly routed to (ingress nginx uses this label as well).

**Comment by fedepaol** [comment]
> @fedepaol unsure if this is related, but the webhook service
> 
> https://github.com/metallb/metallb/blob/2f62620c2b9218f63901f5aaab4d59440f9c2622/charts/metallb/templates/webhooks.yaml#L159
> appears to be missing `metallb.selectorLabels`. If users happen to have other pods deployed to the same namespace with the label `app.kubernetes.io/component: controller`, they'll be mistakenly routed to (ingress nginx uses this label as well).

Oh, that might be one cause. So depending on how the request to the service gets balanced, it might go to the right pod (metallb's controller) or to another one that is not able to serve the request.

 And @jr0dd 's namespace is "networking", so it _might_ be the case. I am gonna fix it anyway, but @jr0dd would you mind checking if you have other pods than the metallb's controller under `networking`?
Also, is there a reason to not isolate metallb in its own namespace? I think we could add this to the guidelines

**Comment by fedepaol** [comment]
> I use the helm chart and have the same issue since yesterday. And metallb stoped working (not assigning L2 IP anymore)
> 
> Going back to 0.12.1 solved this.

@markussiebert  were you getting the same 500 errors when applying resources? If so, can you check what I asked in https://github.com/metallb/metallb/issues/1481#issuecomment-1177234621 ?
Thanks!

**Comment by nick-oconnor** [comment]
> Also, is there a reason to not isolate metallb in its own namespace?

should be fine as a guideline, however I do appreciate it when charts play nicely together in the same namespace (sometimes it's easier architecturally due to dependencies on common resources).

**Comment by MykolaBordakov** [comment]
Good day. 
I have the same issue. 
I tried to install MetalLB by Helm and by Manifest. 
I use this guide : https://kubernetes.github.io/ingress-nginx/deploy/baremetal/#a-pure-software-solution-metallb
I`m useing kube-flannel as a network fabric. 
Maybe there are some specific settings of web-hooks certs or for some other thing ??

**Comment by fedepaol** [comment]
@MykolaBordakov what issue are you facing? Intermittent webhooks?
Can you file a separate issue and give us the logs of the controller pod (and possibly, of the api server) ?

**Comment by MykolaBordakov** [comment]
@fedepaol Thank you for fast response. Full error text is:
Error from server (InternalError): error when creating "kude_fiels/metaLB_conf.yaml": Internal error occurred: failed calling webhook "ipaddresspoolvalidationwebhook.metallb.io": failed to call webhook: Post "https://webhook-service.metallb-system.svc:443/validate-metallb-io-v1beta1-ipaddresspool?timeout=10s": dial tcp 10.103.140.223:443: connect: connection refused
My controllers  pod error  logs: 
{"level":"error","ts":1658933594.284147,"logger":"cert-rotation","msg":"Error updating webhook with certificate","name":"bgppeers.metallb.io","gvk":"apiextensions.k8s.io/v1, Kind=CustomResourceDefinition","error":"Operation cannot be fulfilled on customresourcedefinitions.apiextensions.k8s.io \"bgppeers.metallb.io\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"github.com/open-policy-agent/cert-controller/pkg/rotator.(*ReconcileWH).Reconcile\n\t/go/pkg/mod/github.com/open-policy-agent/cert-controller@v0.3.0/pkg/rotator/rotator.go:642\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:121\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:320\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:234"}
{"level":"error","ts":1658933594.2842596,"msg":"Reconciler error","controller":"cert-rotator","object":{"name":"webhook-server-cert","namespace":"metallb-system"},"namespace":"metallb-system","name":"webhook-server-cert","reconcileID":"79690fd2-8199-4826-a16f-bcd7516cecc9","error":"Operation cannot be fulfilled on customresourcedefinitions.apiextensions.k8s.io \"bgppeers.metallb.io\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:234"}

**Comment by MykolaBordakov** [comment]
@fedepaol Also, i used this command for installetion: 
kubectl apply -f https://raw.githubusercontent.com/metallb/metallb/v0.13.4/config/manifests/metallb-native.yaml

**Comment by fedepaol** [comment]
Please file a separate issue with the information I asked in https://github.com/metallb/metallb/issues/1481#issuecomment-1196825067
Add the whole controller log, not only a part, look for errors in the apiserver.
Finally, try to narrow it to a single cr that you are applying and send the logs only after you manage to generate the failure.

Also, is the webhook always failing or just intermittent?
